crates/storb_validator/src/validator.rs

```diff
@@ -41,6 +61,211 @@ impl Validator {
+
+            // TODO: if we cannot get quic connections of certain miners should we should punish them?
+            let (_, quic_addresses, quic_miner_uids) = get_id_quic_uids(vali_arc).await;
+
+            let num_miners_to_query = std::cmp::min(10, quic_miner_uids.len()); // TODO: use a constant
+            let idxs_to_query: Vec<usize> =
+                (0..quic_miner_uids.len()).choose_multiple(&mut rng, num_miners_to_query);
+
+            for idx in idxs_to_query {
+                let quic_addr = &quic_addresses[idx];
+                let quic_miner_uid = quic_miner_uids[idx];
+                let db = self.scoring_system.write().await.db.clone();
+
+                info!("Challenging miner {quic_miner_uid} with store request");
+                db.conn.lock().await.execute(
+                    "UPDATE miner_stats SET store_attempts = store_attempts + 1 WHERE miner_uid = ?",
+                    [quic_miner_uid],
+                )?;
+
+                let client =
+                    make_client_endpoint(SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), 0))
+                        .map_err(|e| anyhow!(e.to_string()))?;
+                let socket_addr = multiaddr_to_socketaddr(quic_addr)
+                .context("Could not get SocketAddr from Multiaddr. Ensure that the Multiaddr is not missing any components")?;
+
+                let (_, quic_conn) = match create_quic_client(&client, socket_addr).await {
+                    Ok(connection) => (socket_addr, connection),
+                    Err(e) => {
+                        error!("Failed to establish connection with {}: {}", socket_addr, e);
+                        continue; // TODO: punish miner
+                    }
+                };
+
+                let hash_vec = upload_piece_data(&quic_conn, piece.data.clone()).await?; // TODO: error handle
+                let hash = hash_vec.as_bytes_ref();
+
+                info!("Received hash: {:?}", hash);
+                info!("Expected hash: {:?}", piece_hash);
+
+                if hash != piece_hash {
+                    error!("Hash mismatch for piece from miner {quic_miner_uid}");
+                } else {
+                    info!("Received hash for piece from miner {quic_miner_uid}");
+                    // Increase store successes
+                    db.conn.lock().await.execute(
+                    "UPDATE miner_stats SET store_successes = store_successes + 1 WHERE miner_uid = ?",
+                    [quic_miner_uid],
+                )?;
+                    // Increase total successes
+                    db.conn.lock().await.execute(
+                    "UPDATE miner_stats SET total_successes = total_successes + 1 WHERE miner_uid = ?",
+                    [quic_miner_uid],
+                )?;
+                }
+            }
+        }
+
+        info!("Completed synthetic store challenges");
+
+        // ----==== Synthetic retrieval challenges ====----
```

Comment:
replace with something like this
```suggestion
        // Synthetic retrieval challenges
```

Reply:
No.